lib/m2ee/munin.py

```diff
@@ -220,21 +231,135 @@ def get_stats_from_runtime(client, config):
+
+    output_stats = {}
+    for our_memory_type, pool_names in pool_mapping.items():
+        try:
+            total = sum(
+                [memory_pools_dict[pool_name] for pool_name in pool_names]
+            )
+        except KeyError as exc:
+            got_fields = list(memory_pools_dict.keys())
+            required_fields = list(itertools.chain(*pool_mapping.values()))
+            logger.error(
+                "Collecting JVM memory pool stats failed. Memory pool "
+                "output did not match expected output. Needed: %s. Got: %s",
+                required_fields,
+                got_fields,
+            )
+            # TODO: how can we make this information actionable? Should we ask
+            # customers to contact support if in the Mendix Cloud, otherwise
+            # consult GitHub?
+            raise RuntimeError(
+                "Unable to collect JVM memory pool stats."
+            ) from exc
+        output_stats[our_memory_type] = total
+
+    return output_stats
+
+
+def _populate_stats_by_java_version_old(stats, java_version):
+    warnings.warn(
+        "This calculation method is deprecated! Use "
+        "m2ee.munin._populate_stats_by_java_version instead."
+    )
     if "memorypools" in stats["memory"]:
         memorypools = stats["memory"]["memorypools"]
         if java_version == 7:
+            # This branch should never be reached - according to
+            # https://github.com/mendix/m2ee-tools/commit/95738d, MemoryPools
+            # were only added some time after Mendix 6.5, but all versions of
+            # Mendix 6 use Java 8.
+            raise NotImplementedError("This branch should never be reached.")
             stats["memory"]["code"] = memorypools[0]["usage"]
             stats["memory"]["permanent"] = memorypools[4]["usage"]
             stats["memory"]["eden"] = memorypools[1]["usage"]
             stats["memory"]["survivor"] = memorypools[2]["usage"]
             stats["memory"]["tenured"] = memorypools[3]["usage"]
         else:
             stats["memory"]["code"] = memorypools[0]["usage"]
-            stats["memory"]["permanent"] = memorypools[2]["usage"]
+            # In previous versions of this code, the "Compressed Class Space"
+            # pool at index 2 was erroneously used as "Permanent", when the
+            # correct pool is the "Metaspace" pool at index 1.
+            stats["memory"]["permanent"] = memorypools[1]["usage"]
             stats["memory"]["eden"] = memorypools[3]["usage"]
             stats["memory"]["survivor"] = memorypools[4]["usage"]
             stats["memory"]["tenured"] = memorypools[5]["usage"]
     elif java_version >= 8:
+        # This branch should only be reached with a Mendix runtime version
+        # <= 6.5 (i.e. before MemoryPools were added), and one using Java 8,
+        # which seems to include all Mendix 6 runtimes, and some (but not all)
+        # Mendix 5 runtimes.
```

Comment:
@pappachino - this is the remaining case that needs to be re-implemented. This is the case where the original fields were all completely wrong. This is only the case for Mendix 6 and Java 8, without memorypools (The unit tests should make this clear, I hope)